samples/bluetooth/channel_sounding/src/distance_estimation.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+#include <zephyr/bluetooth/cs.h>
+
+#define CS_FREQUENCY_MHZ(ch)    (2402u + 1u * (ch))
+#define CS_FREQUENCY_HZ(ch)     (CS_FREQUENCY_MHZ(ch) * 1000000.0f)
+#define SPEED_OF_LIGHT_M_PER_S  (299792458.0f)
+#define SPEED_OF_LIGHT_NM_PER_S (SPEED_OF_LIGHT_M_PER_S / 1000000000.0f)
+#define PI                      3.14159265358979323846f
+#define MAX_NUM_SAMPLES         256
+
+struct iq_sample_and_channel {
+	bool failed;
+	uint8_t channel;
+	uint8_t antenna_permutation;
+	struct bt_le_cs_iq_sample local_iq_sample;
+	struct bt_le_cs_iq_sample peer_iq_sample;
+};
+
+struct rtt_timing {
+	bool failed;
+	int16_t toa_tod_initiator;
+	int16_t tod_toa_reflector;
+};
+
+static struct iq_sample_and_channel mode_2_data[MAX_NUM_SAMPLES];
+static struct rtt_timing mode_1_data[MAX_NUM_SAMPLES];
+
+struct processing_context {
+	bool local_steps;
+	uint8_t mode_1_data_index;
+	uint8_t mode_2_data_index;
+	uint8_t n_ap;
+	enum bt_conn_le_cs_role role;
+};
+
+static void calc_complex_product(int32_t z_a_real, int32_t z_a_imag, int32_t z_b_real,
+				 int32_t z_b_imag, int32_t *z_out_real, int32_t *z_out_imag)
+{
+	*z_out_real = z_a_real * z_b_real - z_a_imag * z_b_imag;
+	*z_out_imag = z_a_real * z_b_imag + z_a_imag * z_b_real;
+}
+
+static float linear_regression(float *x_values, float *y_values, uint8_t n_samples)
+{
+	if (n_samples == 0) {
+		return 0.0;
+	}
+
+	/* Estimates b in y = a + b x */
+
+	float y_mean = 0.0;
+	float x_mean = 0.0;
+
+	for (uint8_t i = 0; i < n_samples; i++) {
+		y_mean += (y_values[i] - y_mean) / (i + 1);
+		x_mean += (x_values[i] - x_mean) / (i + 1);
+	}
+
+	float b_est_upper = 0.0;
+	float b_est_lower = 0.0;
+
+	for (uint8_t i = 0; i < n_samples; i++) {
+		b_est_upper += (x_values[i] - x_mean) * (y_values[i] - y_mean);
+		b_est_lower += (x_values[i] - x_mean) * (x_values[i] - x_mean);
+	}
+
+	return b_est_upper / b_est_lower;
+}
+
+static void bubblesort_2(float *array1, float *array2, uint16_t len)
+{
+	bool swapped;
+	float temp;
+
+	for (uint16_t i = 0; i < len - 1; i++) {
```

Comment:
can we use qsort() from standard library to avoid O(n^2) complexity?
Not sure how much time we actually spend sorting but I think it can improve performance a bit